### PR TITLE
Canonical ledger caching

### PR DESCRIPTION
**Pull request by Isaac-DeFrain**
This PR rethinks the way we calculate ledgers

- adds `canonical_tip` and `diffs_map` fields to `IndexerState` to enable us to cache ledgers and only apply diffs when needed
- adds `CanonicityStore` trait and corresponding column family to `IndexerStore` for persisting block statuses
- changes `root` and `best_tip` fields to `NodeId`s and adds block accessor functions

Unit tests are not included in this PR! They need to be added

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204865169394584

### CODE REVIEW

**Comment by Isaac-DeFrain** [comment]
I realized that clearing `diffs_map` was wrong. It has been corrected!

**Comment by Isaac-DeFrain** [comment]
Investigating stack overflow! Things are heating up. Time to bust out flamegraph 🔥

**Comment by Isaac-DeFrain** [comment]
As far as I'm concerned, this is G2G!
Adding mainnet ledger tests would be a good addition.
Also, we may want to experiment with different values for `CANONICAL_UPDATE_THRESHOLD` and `PRUNE_INTERVAL_DEFAULT` to find the right balance between ram usage and mainnet ingestion times. At their current values, I'm never seeing more than 2GB ram usage.

File: src/state/mod.rs [hunk 1]
```diff
@@ -97,29 +127,40 @@ impl IndexerState {
...
             transition_frontier_length,
             prune_interval,
+            canonical_update_threshold,
             blocks_processed: 0,
             time: Instant::now(),
             date_time: OffsetDateTime::now_utc(),
         })
     }
 
+    #[allow(clippy::too_many_arguments)]
```
Comment (jenr24G) [inline src/state/mod.rs:153]: lol we'll need to fix this eventually

File: src/store.rs [hunk 2]
```diff
@@ -114,6 +122,47 @@ impl LedgerStore for IndexerStore {
     }
 }
 
+impl CanonicityStore for IndexerStore {
```
Comment (jenr24G) [inline src/store.rs:125]: can we add canonicity to the BlockStore with something like `updateCanonicity`?

Comment (Isaac-DeFrain) [reply src/store.rs:125]: I'm all for that. Would you minding doing it? I really need to focus on CI stuff

Comment (jenr24G) [reply src/store.rs:125]: got it bestie

**Review by jenr24G — APPROVED** [review]

**Pull request merged**